### PR TITLE
Added dynamic table declaration

### PR DESCRIPTION
**Pull request by gleb619**

### CODE REVIEW

File: src/main/java/pl/exsio/querydsl/entityql/Q.java [hunk 1]
```diff
@@ -194,4 +204,12 @@ void addPrimaryKey(List<QEntityColumnMetadata> ids) {
     public PrimaryKey<E> primaryKey() {
         return (PrimaryKey<E>) id;
     }
-}
\ No newline at end of file
+
+    private QEntityScanner qEntityScannerCheck() {
+        if (!(scanner instanceof QEntityScanner)) {
+            throw new UnsupportedScannerException(scanner);
```
Comment (eXsio) [inline src/main/java/pl/exsio/querydsl/entityql/Q.java]: please add String description with the type of operation that won't be supported

File: src/main/java/pl/exsio/querydsl/entityql/entity/scanner/NativeQEntityScanner.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.NamingStrategy;
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.Table;
+import pl.exsio.querydsl.entityql.ex.MissingIdException;
+
+import java.util.List;
+
+/**
+ * QEntityScanner implementation based on Spring Data JDBC.
+ */
+public class NativeQEntityScanner implements TableScanner {
```
Comment (eXsio) [inline src/main/java/pl/exsio/querydsl/entityql/entity/scanner/NativeQEntityScanner.java]: Is it Native though? I'd call it "Dynamic"

File: src/main/java/pl/exsio/querydsl/entityql/QFactory.java [hunk 3]
```diff
@@ -1,7 +1,11 @@
 package pl.exsio.querydsl.entityql;
 
+import com.querydsl.core.Tuple;
 import pl.exsio.querydsl.entityql.entity.metadata.QEntityMetadata;
 import pl.exsio.querydsl.entityql.entity.scanner.QEntityScanner;
+import pl.exsio.querydsl.entityql.entity.scanner.QObjectScanner;
+import pl.exsio.querydsl.entityql.entity.scanner.TableScanner;
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.Table;
```
Comment (eXsio) [inline src/main/java/pl/exsio/querydsl/entityql/QFactory.java]: I would change the package name from "nativesql" to "dynamic". Everything we do here revolves around Native SQL afterall :)

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/dto/Car.groovy [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package pl.exsio.querydsl.entityql.groovy.native_sql.dto
```
Comment (eXsio) [inline src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/dto/Car.groovy]: "native_sql" -> "dynamic"

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    def exportMetadata(DataSource dataSource, String schemaFilter, String tableFilter) {
+        SchemaCrawlerOptions options = SchemaCrawlerOptionsBuilder.builder().toOptions()
+        Catalog catalog = SchemaCrawlerUtility.getCatalog(dataSource.getConnection(), options)
+        def schema = catalog.getSchemas().find { it.getName() =~ schemaFilter }
+        convertTables(catalog, schema, tableFilter)
+    }
+
+    List<pl.exsio.querydsl.entityql.entity.scanner.nativesql.Table> convertTables(Catalog catalog, Schema schema, String tableFilter) {
```
Comment (eXsio) [inline src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy]: please remove fully qualified name

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForTests.groovy [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package pl.exsio.querydsl.entityql.groovy.native_sql.it
+
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.Column
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.Table
+import pl.exsio.querydsl.entityql.groovy.config.enums.by_name.UserTypeByName
+import pl.exsio.querydsl.entityql.groovy.config.enums.by_ordinal.UserTypeByOrdinal
+import spock.lang.Specification
+
+abstract class BaseClassForTests extends Specification {
```
Comment (eXsio) [inline src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForTests.groovy]: please rename to DynamicTestsBase

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+import schemacrawler.schema.Column
+import schemacrawler.schema.Schema
+import schemacrawler.schemacrawler.SchemaCrawlerOptions
+import schemacrawler.schemacrawler.SchemaCrawlerOptionsBuilder
+import schemacrawler.utility.SchemaCrawlerUtility
+import spock.lang.Specification
+
+import javax.sql.DataSource
+
+abstract class BaseClassForDynamicTests extends Specification {
```
Comment (eXsio) [inline src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy]: please rename to DynamicRuntimeTestsBase

File: src/main/java/pl/exsio/querydsl/entityql/type/SimpleMap.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Nullable
+    @Override
+    public <R, C> R accept(Visitor<R, C> v, @Nullable C context) {
+        return v.visit(this, context);
+    }
+
+    @Override
+    public List getArgs() {
```
Comment (eXsio) [inline src/main/java/pl/exsio/querydsl/entityql/type/SimpleMap.java:40]: @SupressWarnings("unchecked")

File: src/main/java/pl/exsio/querydsl/entityql/type/CustomProjections.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+import com.google.common.collect.ImmutableList;
+import com.querydsl.core.types.Expression;
+import com.querydsl.core.types.Path;
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.NamingStrategy;
+import pl.exsio.querydsl.entityql.entity.scanner.nativesql.UnderscoreToCamelStrategy;
+
+import java.util.List;
+
+public final class CustomProjections {
```
Comment (eXsio) [inline src/main/java/pl/exsio/querydsl/entityql/type/CustomProjections.java]: This name is not telling is much. Maybe MapProjections? StringMapProjections? SimpleMapProjections?

**Review by eXsio — CHANGES_REQUESTED** [review]
Hello, thank you for the PR.

Please review and fix the comments (mainly naming). 
Please review the CodeCov report and make sure that the overall code coverage is not decreased (currently is down almost 5%).
Please make sure all the dependencies are there (I've recently removed Guava and I can see that your code depends on it)

File: src/main/java/pl/exsio/querydsl/entityql/QFactory.java (hunk 3 above)
Comment (gleb619) [reply src/main/java/pl/exsio/querydsl/entityql/QFactory.java]: Couldn't think of anything original)
In progress....

File: src/main/java/pl/exsio/querydsl/entityql/type/CustomProjections.java (hunk 9 above)
Comment (eXsio) [reply src/main/java/pl/exsio/querydsl/entityql/type/CustomProjections.java]: Or maybe RuntimeProjections? If the feature you're adding should be used only when the schema is not known during compilation time, then it should be "Runtime"

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForTests.groovy (hunk 6 above)
Comment (eXsio) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForTests.groovy]: correction, rename to RuntimeFixedTestBase

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy (hunk 7 above)
Comment (eXsio) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy]: correction, rename to RuntimeDBTestBase

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/dto/Car.groovy (hunk 4 above)
Comment (eXsio) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/dto/Car.groovy]: correction: "native_sql" -> "runtime"

File: src/main/java/pl/exsio/querydsl/entityql/QFactory.java (hunk 3 above)
Comment (eXsio) [reply src/main/java/pl/exsio/querydsl/entityql/QFactory.java]: correction: "nativesql" -> "runtime"

File: src/main/java/pl/exsio/querydsl/entityql/entity/scanner/NativeQEntityScanner.java (hunk 2 above)
Comment (eXsio) [reply src/main/java/pl/exsio/querydsl/entityql/entity/scanner/NativeQEntityScanner.java]: correction: "NativeQEntityScanner" -> "RuntimeTableScanner"

**Comment by eXsio** [comment]
Ok, couple more naming changes. I think naming should be consistent. As we want to add handing of Runtime schema (without knowing it during compilation time), then let's add "Runtime" to all significant names.

File: src/main/java/pl/exsio/querydsl/entityql/Q.java (hunk 1 above)
Comment (gleb619) [reply src/main/java/pl/exsio/querydsl/entityql/Q.java]: Added description

File: src/main/java/pl/exsio/querydsl/entityql/entity/scanner/NativeQEntityScanner.java (hunk 2 above)
Comment (gleb619) [reply src/main/java/pl/exsio/querydsl/entityql/entity/scanner/NativeQEntityScanner.java]: renamed to RuntimeTableScanner

File: src/main/java/pl/exsio/querydsl/entityql/QFactory.java (hunk 3 above)
Comment (gleb619) [reply src/main/java/pl/exsio/querydsl/entityql/QFactory.java]: renamed

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/dto/Car.groovy (hunk 4 above)
Comment (gleb619) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/dto/Car.groovy]: renamed

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy (hunk 5 above)
Comment (gleb619) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy]: removed fqn

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForTests.groovy (hunk 6 above)
Comment (gleb619) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForTests.groovy]: renamed to RuntimeFixedTestBase

File: src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy (hunk 7 above)
Comment (gleb619) [reply src/test/groovy/pl/exsio/querydsl/entityql/groovy/native_sql/it/BaseClassForDynamicTests.groovy]: renamed to RuntimeDBTestBase

File: src/main/java/pl/exsio/querydsl/entityql/type/SimpleMap.java (hunk 8 above)
Comment (gleb619) [reply src/main/java/pl/exsio/querydsl/entityql/type/SimpleMap.java:40]: added

File: src/main/java/pl/exsio/querydsl/entityql/type/CustomProjections.java (hunk 9 above)
Comment (gleb619) [reply src/main/java/pl/exsio/querydsl/entityql/type/CustomProjections.java]: I don't think Projections has anything to do with runtime, renamed it SimpleMapProjections.

**Comment by eXsio** [comment]
couple of renames left and we'll be able to merge :)

**Comment by gleb619** [comment]
I don't know what to rename

**Comment by eXsio** [comment]
1. Remove TableScanner interface - make RuntimeTableScanner implement  QObjectScanner<Table> directly
2. Rename "Column" -> "RuntimeColumn"
3. Rename "Table" -> "RuntimeTable"
4. Rename "SimpleMap" -> "RuntimeMap"
5. Rename "CustomProjections" -> "RuntimeProjections"
6. Rename "NamingStrategy" -> "RuntimeNamingStrategy"
7. Rename "SimpleColumn" -> "SimpleRuntimeColumn"
8. Rename "SimpleTable" -> "SimpleRuntimeTable"

**Comment by gleb619** [comment]
Well, i can do it tomorrow if you want. But this PR already merged)

**Pull request merged**